cql/src/main/java/org/apache/cassandra/stargate/transport/internal/PlainTextTokenSaslNegotiator.java

```diff
@@ -0,0 +1,113 @@
+package org.apache.cassandra.stargate.transport.internal;
+
+import io.stargate.auth.AuthenticationService;
+import io.stargate.auth.StoredCredentials;
+import io.stargate.db.AuthenticatedUser;
+import io.stargate.db.Authenticator;
+import io.stargate.db.Persistence;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import org.apache.cassandra.stargate.exceptions.AuthenticationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class PlainTextTokenSaslNegotiator implements Authenticator.SaslNegotiator {
+  private static final Logger logger = LoggerFactory.getLogger(PlainTextTokenSaslNegotiator.class);
+
+  static final String TOKEN_USERNAME = System.getProperty("stargate.cql_token_username", "token");
+  static final int TOKEN_MAX_LENGTH =
+      Integer.parseInt(System.getProperty("stargate.cql_token_max_length", "36"));
+
+  static final byte NUL = 0;
+
+  private final Persistence persistence;
+  private final AuthenticationService authentication;
+  private final Authenticator.SaslNegotiator wrapped;
+  private StoredCredentials credentials;
+  private String username;
+  private String password;
+
+  PlainTextTokenSaslNegotiator(
+      Authenticator.SaslNegotiator wrapped,
+      Persistence persistence,
+      AuthenticationService authentication) {
+    this.persistence = persistence;
+    this.authentication = authentication;
+    this.wrapped = wrapped;
+  }
+
+  @Override
+  public byte[] evaluateResponse(byte[] clientResponse) throws AuthenticationException {
+    if (attemptTokenAuthentication(clientResponse)) return null;
+    return wrapped.evaluateResponse(clientResponse);
+  }
+
+  @Override
+  public boolean isComplete() {
+    return credentials != null || wrapped.isComplete();
+  }
+
+  @Override
+  public AuthenticatedUser<?> getAuthenticatedUser() throws AuthenticationException {
+    if (credentials != null) return persistence.newAuthenticatedUser(credentials.getRoleName());
+    else return wrapped.getAuthenticatedUser();
+  }
+
+  private boolean attemptTokenAuthentication(byte[] clientResponse) {
+    try {
+      decodeCredentials(clientResponse);
```

Comment:
I find this design a bit misleading. It seems that `decodeCredentials` is called every time the `attemptTokenAuthentication` is called, but it sets the `password` and `username` fields on this class. Those fields are no used outside of this context, so I think that this should be reworked to stateless logic. The `decodeCredentials` should return username/password (maybe wrapped in some light class) instead of setting class fields. It will make logic simpler and easier to test. It would make sense to have class fields if the results of the `decodeCredentials` will be memoized.

Reply:
Yes, that makes sense.